File: 2018/04_ReposeRecord/shift.py
# ======================================================================
# Repose Record
#   Advent of Code 2018 Day 04 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                             s h i f t . p y
# ======================================================================
"Shift for Repose Record problem for day 04 of Advent of Code"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import Counter
import re

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
YMDHM = r'^\[1518-([0-9][0-9])-([0-9][0-9]) ([0-9][0-9]):([0-9][0-9])]'
GUARD = re.compile(YMDHM + r' Guard #([0-9]+) begins shift$')
FALLS = re.compile(YMDHM + r' falls asleep$')
WAKES = re.compile(YMDHM + r' wakes up$')

#               JAN FEB MAR APR MAY JUN JUL AUG SEP OCT NOV DEC
DAYSMONTH = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]



# ======================================================================
#                                                                  Shift
# ======================================================================


class Shift():
    """Object representing a single row of fabric"""

    def __init__(self,                          # pylint: disable=R0913
                 number=0, month=1, day=1, hour=0, text=None):

        # 1. Set the values
        self.number = number
        self.month = month
        self.day = day
        self.falls = 0
        self.wakes = 0
        self.minutes = Counter()

        # 2. Process text (if any)
        if text:
            match = GUARD.match(text)
            assert match is not None
            self.month = int(match.group(1))
            self.day = int(match.group(2))
            hour = int(match.group(3))
            self.number = int(match.group(5))

        # 3. Tweak month and day (if needed)
        if hour == 23:
            self.day += 1
            if self.day > DAYSMONTH[self.month]:
                self.day = 1
                self.month += 1
                if self.month > 12:
                    self.month = 1

        # 4. Postconditions
        assert self.month >= 1 and month <= 12
        assert self.day >= 1 and day <= 31
        assert self.falls == 0
        assert self.wakes == 0

    def add_fall(self, falls=0, text=None):
        "Record time of falling asleep"

        # 0. Preconditions
        assert self.falls == 0
        assert self.wakes == 0

        # 1. Save the value
        self.falls = falls

        # 2. Process text (if any)
        if text:
            match = FALLS.match(text)
            assert match is not None
            self.month = int(match.group(1))
            self.day = int(match.group(2))
            hour = int(match.group(3))
            assert hour == 0
            self.falls = int(match.group(4))

        # 9. Postconditions
        assert self.falls >= 0
        assert self.wakes == 0

    def add_wake(self, wakes=0, text=None):
        "Record time of falling asleep"

        # 0. Preconditions
        assert self.falls >= 0

        # 1. Save the value
        self.wakes = wakes

        # 2. Process text (if any)
        if text:
            match = WAKES.match(text)
            assert match is not None
            self.month = int(match.group(1))
            self.day = int(match.group(2))
            hour = int(match.group(3))
            assert hour == 0
            self.wakes = int(match.group(4))

        # 3. Make claims to squares in this row
        assert self.falls < self.wakes
        self.minutes.update(range(self.falls, self.wakes))

        # 4. Reset for next time
        self.falls = 0
        self.wakes = 0

    def asleep(self):
        "Total minutes asleep this shift"

        return sum(self.minutes.values())


    def often(self):
        "Return the minute and times most often asleep"

        if len(self.minutes) > 0:
            return self.minutes.most_common(1)[0]
        return (0, 0)

    def __str__(self):

        view = []
        for minute in range(60):
            if minute in self.minutes:
                view.append('#')
            else:
                view.append('.')
        return "%02d-%02d  #%04d  %s" % (self.month, self.day, self.number, ''.join(view))


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        s h i f t . p y                         end
# ======================================================================

File: 2018/04_ReposeRecord/shift.go
// ======================================================================
// Repose Record
//   Advent of Code 2018 Day 04 -- Eric Wastl -- https://adventofcode.com
//
// lua implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         s h i f t . g o
// ======================================================================
// Shift for the Advent of Code 2018 Day 04 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"regexp"
	"strconv"
)

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const (
	EXAMPLE_TEXT = "[1518-11-01 00:00] Guard #10 begins shift"
)

var (
	shiftExp = regexp.MustCompile("\\d+")
)

// ======================================================================
//                                                                  Shift
// ======================================================================

type Shift struct {
	Part2  bool
	Text   string
	Date   string
	Asleep []bool
}

func NewShift(part2 bool, text string) *Shift {
	// Shift Object for Repose Record

	// 1. Set the initial values
	p := new(Shift)
	p.Part2 = part2
	p.Text = text
	p.Date = "unknown"
	p.Asleep = make([]bool, 60)

	// 2. Set the date (if we can)
	if len(text) > 12 {
		p.Date = text[1:11]
	}

	// 3. Return the shift
	return p
}

func (p *Shift) FallsAlseep(text string) {
	// Add sleep time to the shift

	// 1. Parse the date and time
	parts := shiftExp.FindAllString(text, -1)
	if len(parts) != 5 {
		fmt.Printf("regexp failed for line '%s'\n", text)
		return
	}

	// 2. Get the minute the guard fell asleep
	start, _ := strconv.Atoi(parts[4])

	// 3. Mark when the guard is asleep
	for minute := start; minute < 60; minute++ {
		p.Asleep[minute] = true
	}

	// 4. Save the date (if case shift started before midnight)
	p.Date = text[1:11]
}

func (p *Shift) WakesUp(text string) {
	// Remove sleep time from the shift

	// 1. Parse the date and time
	parts := shiftExp.FindAllString(text, -1)
	if len(parts) != 5 {
		fmt.Printf("regexp failed for line '%s'\n", text)
		return
	}

	// 2. Get the minute the guard woke up
	start, _ := strconv.Atoi(parts[4])

	// 3. Mark when the guard is awake
	for minute := start; minute < 60; minute++ {
		p.Asleep[minute] = false
	}
}

func (p *Shift) MinutesAsleep() int {
	// Return the number of minutes a guard has been asleep this shift

	// 1. Start with nothing
	minutes := 0

	// 2. Loop for all of the minutes of the shift
	for _, minute := range p.Asleep {

		// 3. If sleeping for this minute, add to total
		if minute {
			minutes++
		}
	}

	// 4. Return the number of minutes asleep
	return minutes
}

// ======================================================================
// end                      s h i f t . g o                     end
// ======================================================================
